dkg-gadget/src/testing.rs

```diff
@@ -0,0 +1,1210 @@
+
+	pub fn send(&self, t: T) {
+		let mut lock = self.inner.write();
+		assert!(!lock.is_empty());
+		// receiver will naturally drop when no longer used.
+		lock.retain(|tx| tx.unbounded_send(t.clone()).is_ok())
+	}
+}
+
+impl BlockchainEvents<TestBlock> for TestBackend {
+	fn finality_notification_stream(&self) -> sc_client_api::FinalityNotifications<TestBlock> {
+		self.inner.finality_stream.subscribe()
+	}
+
+	fn import_notification_stream(&self) -> sc_client_api::ImportNotifications<TestBlock> {
+		self.inner.import_stream.subscribe()
+	}
+
+	fn storage_changes_notification_stream(
+		&self,
+		_filter_keys: Option<&[sc_client_api::StorageKey]>,
+		_child_filter_keys: Option<
+			&[(sc_client_api::StorageKey, Option<Vec<sc_client_api::StorageKey>>)],
+		>,
+	) -> sp_blockchain::Result<sc_client_api::StorageEventStream<<TestBlock as BlockT>::Hash>> {
+		todo!()
+	}
+}
+
+impl sc_client_api::Backend<TestBlock> for TestBackend {
+	type BlockImportOperation = DummyStateBackend;
+	type Blockchain = sc_client_api::in_mem::Blockchain<TestBlock>;
+	type State = DummyStateBackend;
+	type OffchainStorage = InMemOffchainStorage;
+
+	fn pin_block(
+		&self,
+		_: <TestBlock as sp_api::BlockT>::Hash,
+	) -> Result<(), sp_blockchain::Error> {
+		todo!()
+	}
+	fn unpin_block(&self, _: <TestBlock as sp_api::BlockT>::Hash) {
+		todo!()
+	}
+
+	fn begin_operation(&self) -> sp_blockchain::Result<Self::BlockImportOperation> {
+		todo!()
+	}
+
+	fn begin_state_operation(
+		&self,
+		_operation: &mut Self::BlockImportOperation,
+		_block: <TestBlock as BlockT>::Hash,
+	) -> sp_blockchain::Result<()> {
+		todo!()
+	}
+
+	fn commit_operation(
+		&self,
+		_transaction: Self::BlockImportOperation,
+	) -> sp_blockchain::Result<()> {
+		todo!()
+	}
+
+	fn finalize_block(
+		&self,
+		_hash: <TestBlock as BlockT>::Hash,
+		_justification: Option<sp_runtime::Justification>,
+	) -> sp_blockchain::Result<()> {
+		todo!()
+	}
+
+	fn append_justification(
+		&self,
+		_hash: <TestBlock as BlockT>::Hash,
+		_justification: sp_runtime::Justification,
+	) -> sp_blockchain::Result<()> {
+		todo!()
+	}
+
+	fn blockchain(&self) -> &Self::Blockchain {
+		todo!()
+	}
+
+	fn usage_info(&self) -> Option<sc_client_api::UsageInfo> {
+		todo!()
+	}
+
+	fn offchain_storage(&self) -> Option<Self::OffchainStorage> {
+		Some(self.inner.offchain_storage.clone())
+	}
+
+	fn state_at(&self, _hash: <TestBlock as BlockT>::Hash) -> sp_blockchain::Result<Self::State> {
+		todo!()
+	}
+
+	fn revert(
+		&self,
+		_n: sp_api::NumberFor<TestBlock>,
+		_revert_finalized: bool,
+	) -> sp_blockchain::Result<(
+		sp_api::NumberFor<TestBlock>,
+		std::collections::HashSet<<TestBlock as BlockT>::Hash>,
+	)> {
+		todo!()
+	}
+
+	fn remove_leaf_block(&self, _hash: <TestBlock as BlockT>::Hash) -> sp_blockchain::Result<()> {
+		todo!()
+	}
+
+	fn get_import_lock(&self) -> &RwLock<()> {
+		todo!()
+	}
+
+	fn requires_full_sync(&self) -> bool {
+		todo!()
+	}
+}
+
+impl HeaderBackend<TestBlock> for TestBackend {
```

Comment:
Do these need to be implemented or stubs are fine?

Reply:
I tried stubs at first, but, this doesn't work since this trait implementation is required when we have a `TestBlock`.